Add ChaCha and Salsa intrinsics.

TimothyMakkison opened the pull request:
Added @macaba intrinsics implementation #58 for ChaCha20. Salsa20 has since been added to NaCl, so I created `Salsa20BaseIntrinsics` based on macaba's work. At minimum both got a 4x speed up.

- Added ChaCha20 & Salsa20.
- Added benchmarks for Salsa20 and XSalsa20.
- Added checks for intrinsics support and failovers.
- Added `VariableLengthCiphers` tests, to ensure that the intrinsics methods work when using length specific methods.


This pr will require a way of testing with and without intrinsics, I'm not sure how to do this but I think `COMPlus_EnableAVX` is related? 

A future update could speed up `ProcessKeyStreamblock` by using the `ProcessStream` method.
``` C#
private static byte[] _zeros = new byte[BLOCK_SIZE_IN_BYTES];

public override void ProcessKeyStreamBlock(ReadOnlySpan<byte> nonce, int counter, Span<byte> block)
{

    if (block.Length != BLOCK_SIZE_IN_BYTES)
        throw new CryptographicException($"The key stream block length is not valid. The length in bytes must be {BLOCK_SIZE_IN_BYTES}.");
#if INTRINSICS
    if (Sse2.IsSupported)
    {
        ProcessStream(nonce, block, _zeros, counter);
        return;
    }
...
```

macaba committed:
Initial integration of intrinsics

macaba committed:
Update Program.cs

macaba committed:
Removed 5.0 target to allow CI to build

TimothyMakkison committed:
Resolve conflicts

TimothyMakkison committed:
Remove treat warning as errors

TimothyMakkison committed:
Alter Dencrypt test

TimothyMakkison committed:
Add Salsa20 and XSalsa20 benchmarks

TimothyMakkison committed:
Added back .net 6 support

TimothyMakkison committed:
Parameterise Salsa20TestVectors

TimothyMakkison committed:
Added Intrinsics check

TimothyMakkison committed:
Added intrinsics support for Salsa20 (64 <= is broken)

TimothyMakkison committed:
Fix Salsa20 intrinsics for bytes >= 64

TimothyMakkison committed:
Merge benchmark updates

TimothyMakkison committed:
Fix benchmark

TimothyMakkison committed:
Use transpose rotate transpose trick - 10-20% faster

TimothyMakkison committed:
Add Transpose method

TimothyMakkison committed:
Revert "Add Transpose method"

This reverts commit cbf3778b81eed7231cc633a48bf641061ea67004.

TimothyMakkison committed:
Add transpose method

TimothyMakkison committed:
Fix XSalsa benchmark

TimothyMakkison committed:
Code cleanup

TimothyMakkison committed:
Correct nonce size

TimothyMakkison committed:
Inline pre processor variable

TimothyMakkison committed:
Add variable length test for ChaCha and Salsa

daviddesmet commented:
Thanks for the contribution and for adding Salsa20 intrinsics too!
I wonder if we can reduce the method complexity of those intrinsics ones by breaking them apart. Afaik, the initial code was grabbed from another repository which I currently don't recall.

There's also XSalsa20 over there but I haven't promoted those to stable since I felt it needed more tests. The target was to also support XSalsa20Poly1305.

Regarding testing of intrinsics turned on and off, yes, it was something around `COMPlus_EnableAVX`, I'm not sure since I looked at this a while ago. For the CI, we might need to add tests for running with and without intrinsics I guess.

Btw, do you have some benchmark numbers you can share to see the perf increase?

I might need to take a look at why the Azure CI failed with Windows only.

TimothyMakkison committed:
Refactor ChaCha20BaseIntrinsics

TimothyMakkison committed:
Refactor Salsa20BaseIntrinsics

TimothyMakkison committed:
Added HChaCha and HSalsa

TimothyMakkison commented:
> I wonder if we can reduce the method complexity of those intrinsics ones by breaking them apart.
> 

I've split Intrinsics up by chunk size and altered `ChaCha64` and `Salsa64` to add `HChaCha`/`HSalsa` support. Aside from that I haven't changed the other methods. I guess I could extract some of the Xor/Store/Unpack methods and give them nicer names. I'll add `ProcessKeyStreamIntrinsics`, the changes to the 64 chunk methods should make it painless.

> There's also XSalsa20 over there but I haven't promoted those to stable since I felt it needed more tests. The target was to also support XSalsa20Poly1305.
> 
> Regarding testing of intrinsics turned on and off, yes, it was something around `COMPlus_EnableAVX`, I'm not sure since I looked at this a while ago. For the CI, we might need to add tests for running with and without intrinsics I guess.

I'll take a look, I've never used env variables in unit tests before, hopefully Google will help.

> 
> Btw, do you have some benchmark numbers you can share to see the perf increase?
> 
My numbers are from some benchmarks I did yesterday. Sorry about the quality, my laptops about to die 😅

### Salsa
![image](https://user-images.githubusercontent.com/49349513/195461071-926afe26-41fd-4203-a7f4-b71b08bd253b.png)

### ChaCha
![image](https://user-images.githubusercontent.com/49349513/195461306-0ad9b553-187e-449d-8d4a-fb8ef9e7124c.png)

daviddesmet commented:
> I've split Intrinsics up by chunk size and altered `ChaCha64` and `Salsa64` to add `HChaCha`/`HSalsa` support. Aside from that I haven't changed the other methods. I guess I could extract some of the Xor/Store/Unpack methods and give them nicer names. I'll add `ProcessKeyStreamIntrinsics`, the changes to the 64 chunk methods should make it painless.

Sounds good, thanks a lot for the effort you are putting into this.

> I'll take a look, I've never used env variables in unit tests before, hopefully Google will help.

If you run the benchmarks from Powershell or cmd via `dotnet run`, you can set the variable using `set VARIABLE=VALUE` before the `dotnet` command.

> My numbers are from some benchmarks I did yesterday. Sorry about the quality, my laptops about to die 😅
> 
> ### Salsa
> ![image](https://user-images.githubusercontent.com/49349513/195461071-926afe26-41fd-4203-a7f4-b71b08bd253b.png)
> 
> ### ChaCha
> ![image](https://user-images.githubusercontent.com/49349513/195461306-0ad9b553-187e-449d-8d4a-fb8ef9e7124c.png)

Thanks for sharing those, the numbers look nice :)

TimothyMakkison committed:
Added little endian check

TimothyMakkison committed:
Added intrinsics process stream

TimothyMakkison committed:
Added Salsa ProcessKeyStreamBlock test

TimothyMakkison committed:
Minor HSalsa & KeyStream api changes

TimothyMakkison committed:
Remove pre processor per method system checks. Instead using either a normal Core or IntrinsicsCore if supported

TimothyMakkison committed:
Use ChaChaCore or ChaChaIntrinsics instead of pre processor checks

TimothyMakkison committed:
Delete Snuffle pre processor checks

TimothyMakkison committed:
Delete Snuffle pre processor checks

TimothyMakkison committed:
Resolve

TimothyMakkison committed:
Fix process stackoverflow bug

TimothyMakkison commented:
I've experimented with a [refactor](https://github.com/TimothyMakkison/NaCl.Core/tree/experimental_intrinsics_refactor) that treats `ChaCha20Base` and `Salsa20Base` as a proxy for the internal `...Core` and `...CoreItrinsics` methods. It removes the intrinsics checks and preprocessor spam but arguably makes the code harder to read, I don't think I changed any externally visible apis.

I'm planning on creating some general tests for intrinsics code and relying on the CI with different flags enabled to do a thorough check. Ideally I'd configure XUnit to run the tests/test classes again with different flags but it doesn't seem supported 🤷

TimothyMakkison committed:
Fix incorrect system checks

TimothyMakkison committed:
Fix incorrect system checks

TimothyMakkison committed:
Resolve

TimothyMakkison committed:
Rewrite HSalsa to use only Sse2

TimothyMakkison committed:
Minor refactoring, add comments

TimothyMakkison committed:
Refactor BaseIntrinsics to use pointers

TimothyMakkison committed:
Resolve conflicts

TimothyMakkison committed:
Move Core files into folders

TimothyMakkison committed:
Fix process methods

TimothyMakkison committed:
Added powershell test file, runs tests with various simd modes enabled/disabled

TimothyMakkison committed:
Update core namespaces

TimothyMakkison committed:
Added Intrinsic/Scalar tests

TimothyMakkison committed:
Updated Intrinsics test powershell script

TimothyMakkison committed:
Code cleanup and formatting changes

daviddesmet commented on src/NaCl.Core/Base/ChaCha20Base.cs:
```diff
@@ -2,22 +2,37 @@
     /// <seealso cref="NaCl.Core.Base.Snuffle" />
     public abstract class ChaCha20Base : Snuffle
     {
+        readonly IChaCha20Core _chaCha20Core;
```
Could you make this access modifier implicit?

daviddesmet commented on src/NaCl.Core/Base/ChaCha20Base.cs:
```diff
@@ -2,22 +2,37 @@
-        protected ChaCha20Base(ReadOnlyMemory<byte> key, int initialCounter) : base(key, initialCounter) { }
+        protected ChaCha20Base(ReadOnlyMemory<byte> key, int initialCounter) : base(key, initialCounter)
+        {
+#if INTRINSICS
```
I wonder if we do need the preprocessor directive since we are checking if the hardware supports it

daviddesmet commented on src/NaCl.Core/Base/ChaCha20BaseIntrinsics.cs:
```diff
@@ -0,0 +1,69 @@
+﻿#if INTRINSICS
```
I wonder if we do need the preprocessor directive since we are checking if the hardware supports it

daviddesmet commented on src/NaCl.Core/Base/ChaChaCore/ChaCha20CoreIntrinsics.cs:
```diff
@@ -0,0 +1,54 @@
+﻿#if INTRINSICS
```
I wonder if we do need the preprocessor directive since we are checking if the hardware supports it

daviddesmet commented on src/NaCl.Core/Base/ChaChaIntrinsics/ChaCha256.cs:
```diff
@@ -0,0 +1,174 @@
+﻿#if INTRINSICS
```
I wonder if we do need the preprocessor directive since we are checking if the hardware supports it

daviddesmet commented on src/NaCl.Core/Base/ChaChaIntrinsics/ChaCha512.cs:
```diff
@@ -0,0 +1,244 @@
+﻿#if INTRINSICS
```
I wonder if we do need the preprocessor directive since we are checking if the hardware supports it

daviddesmet commented on src/NaCl.Core/Base/ChaChaIntrinsics/ChaCha64.cs:
```diff
@@ -0,0 +1,200 @@
+﻿#if INTRINSICS
```
I wonder if we do need the preprocessor directive since we are checking if the hardware supports it

daviddesmet commented on src/NaCl.Core/Base/Salsa20Base.cs:
```diff
@@ -2,22 +2,37 @@
     /// <seealso cref="NaCl.Core.Base.Snuffle" />
     public abstract class Salsa20Base : Snuffle
     {
+        readonly ISalsa20Core _salsa20Core;
```
Could you make this access modifier implicit?

daviddesmet commented on src/NaCl.Core/Base/Salsa20Base.cs:
```diff
@@ -29,58 +44,21 @@ public abstract class Salsa20Base : Snuffle
         /// <param name="nonce">The nonce.</param>
         /// <param name="counter">The counter.</param>
-        protected abstract void SetInitialState(Span<uint> state, ReadOnlySpan<byte> nonce, int counter);
+        internal protected abstract void SetInitialState(Span<uint> state, ReadOnlySpan<byte> nonce, int counter);
```
Something minor... the access modifier should be `protected internal`, just need to fix the ordering.

daviddesmet commented on src/NaCl.Core/Base/Salsa20Base.cs:
```diff
@@ -2,22 +2,37 @@
-        protected Salsa20Base(ReadOnlyMemory<byte> key, int initialCounter) : base(key, initialCounter) { }
+        protected Salsa20Base(ReadOnlyMemory<byte> key, int initialCounter) : base(key, initialCounter)
+        {
+#if INTRINSICS
```
I wonder if we do need the preprocessor directive since we are checking if the hardware supports it

daviddesmet commented on src/NaCl.Core/Base/Salsa20Base.cs:
```diff
@@ -112,7 +90,7 @@ public void HSalsa20InitialState(Span<uint> state, ReadOnlySpan<byte> nonce)
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        protected static void ShuffleState(Span<uint> state)
+        internal protected static void ShuffleState(Span<uint> state)
```
Something minor... the access modifier should be `protected internal`, just need to fix the ordering.

daviddesmet commented on src/NaCl.Core/Base/Salsa20BaseIntrinsics.cs:
```diff
@@ -0,0 +1,69 @@
+﻿#if INTRINSICS
```
I wonder if we do need the preprocessor directive since we are checking if the hardware supports it

daviddesmet commented on src/NaCl.Core/Base/SalsaCore/Salsa20CoreIntrinsics.cs:
```diff
@@ -0,0 +1,57 @@
+﻿#if INTRINSICS
```
I wonder if we do need the preprocessor directive since we are checking if the hardware supports it

daviddesmet commented on src/NaCl.Core/Base/SalsaCore/Salsa20CoreIntrinsics.cs:
```diff
@@ -0,0 +1,57 @@
+
+internal class Salsa20CoreIntrinsics : ISalsa20Core
+{
+    const int BLOCK_SIZE_IN_BYTES = Snuffle.BLOCK_SIZE_IN_BYTES;
+    const int BLOCK_SIZE_IN_INTS = Snuffle.BLOCK_SIZE_IN_INTS;
```
Could you make this access modifier implicit?

daviddesmet commented on src/NaCl.Core/Base/SalsaIntrinsics/Salsa256.cs:
```diff
@@ -0,0 +1,177 @@
+﻿#if INTRINSICS
```
I wonder if we do need the preprocessor directive since we are checking if the hardware supports it

daviddesmet commented on src/NaCl.Core/Base/SalsaIntrinsics/Salsa512.cs:
```diff
@@ -0,0 +1,243 @@
+﻿#if INTRINSICS
```
I wonder if we do need the preprocessor directive since we are checking if the hardware supports it

daviddesmet commented on src/NaCl.Core/Base/SalsaIntrinsics/Salsa64.cs:
```diff
@@ -0,0 +1,224 @@
+﻿#if INTRINSICS
```
I wonder if we do need the preprocessor directive since we are checking if the hardware supports it

daviddesmet commented on src/NaCl.Core/Base/Snuffle.cs:
```diff
@@ -22,13 +22,13 @@ public abstract class Snuffle
 
         protected readonly ReadOnlyMemory<byte> Key;
-        protected readonly int InitialCounter;
+        internal protected readonly int InitialCounter;
```
Something minor... the access modifier should be `protected internal`, just need to fix the ordering.

daviddesmet commented on src/NaCl.Core/NaCl.Core.csproj:
```diff
@@ -5,7 +5,6 @@
     <TargetFrameworks Condition="'$(OS)' != 'Unix'">netstandard2.0;netstandard2.1;netcoreapp3.1;net45;net48;net5.0;net6.0</TargetFrameworks>
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
     <LangVersion>latest</LangVersion>
-    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
```
Would be nice to leave this on.

daviddesmet commented on src/NaCl.Core/Salsa20.cs:
```diff
@@ -23,7 +23,7 @@ public class Salsa20 : Salsa20Base
 
         /// <inheritdoc />
-        protected override void SetInitialState(Span<uint> state, ReadOnlySpan<byte> nonce, int counter)
+        internal protected override void SetInitialState(Span<uint> state, ReadOnlySpan<byte> nonce, int counter)
```
Something minor... the access modifier should be `protected internal`, just need to fix the ordering.

daviddesmet commented on src/NaCl.Core/XSalsa20.cs:
```diff
@@ -23,7 +23,7 @@ public class XSalsa20 : Salsa20Base
 
         /// <inheritdoc />
-        protected override void SetInitialState(Span<uint> state, ReadOnlySpan<byte> nonce, int counter)
+        internal protected override void SetInitialState(Span<uint> state, ReadOnlySpan<byte> nonce, int counter)
```
Something minor... the access modifier should be `protected internal`, just need to fix the ordering.

daviddesmet commented on test/NaCl.Core.Benchmarks/ChaCha20Benchmark.cs:
```diff
@@ -1,86 +1,90 @@
-        //public byte[] Decrypt(byte[] ciphertext) => cipher.Decrypt(ciphertext);
-    }
-}
+﻿namespace NaCl.Core.Benchmarks
```
We could switch to scoped namespace declaration here since we already using it on the other files =)

daviddesmet commented on test/NaCl.Core.Benchmarks/Program.cs:
```diff
@@ -8,14 +8,21 @@ class Program
     {
         static void Main(string[] args)
         {
+            var c = new ChaCha20Benchmark();
+            c.Size = 10_000;
+            c.Setup();
+            c.Encrypt();
```
Do we need to call the benchmark like this? Maybe it was some test code that somehow got into here?

daviddesmet replied:
Do we need to call the benchmark like this? Maybe it was some test code that somehow got into here?

daviddesmet commented on test/NaCl.Core.Benchmarks/Salsa20Benchmark.cs:
```diff
@@ -0,0 +1,83 @@
+﻿namespace NaCl.Core.Benchmarks
```
We could switch to scoped namespace declaration here since we already using it on the other files =)

daviddesmet commented on test/NaCl.Core.Benchmarks/XSalsa20Benchmark .cs:
```diff
@@ -0,0 +1,81 @@
+﻿namespace NaCl.Core.Benchmarks
```
We could switch to scoped namespace declaration here since we already using it on the other files =)

daviddesmet commented on test/NaCl.Core.SimdTests/ChaCha20IntrinsicsTests .cs:
```diff
@@ -0,0 +1,347 @@
+﻿namespace NaCl.Core.SimdTests
```
We could switch to scoped namespace declaration here since we already using it on the other files =)

daviddesmet commented on test/NaCl.Core.SimdTests/ChaCha20IntrinsicsTests .cs:
```diff
@@ -0,0 +1,347 @@
+    using Internal;
+    using Vectors;
+    using NaCl.Core.Base.ChaChaCore;
+    using NaCl.Core.SimdTests.Vectors;
```
I don't think this using is needed

daviddesmet commented on test/NaCl.Core.SimdTests/ChaCha20ScalarTests.cs:
```diff
@@ -0,0 +1,347 @@
+    using Internal;
+    using Vectors;
+    using NaCl.Core.Base.ChaChaCore;
+    using NaCl.Core.SimdTests.Vectors;
```
I don't think this using is needed since above it you are already specifying `using Vectors;`

daviddesmet commented on test/NaCl.Core.SimdTests/Salsa20IntrinsicsTests.cs:
```diff
@@ -0,0 +1,433 @@
+    using Vectors;
+    using System.Linq;
+    using NaCl.Core.Base.SalsaCore;
+    using NaCl.Core.SimdTests.Vectors;
```
I don't think this `using` is needed since above it you are already specifying `using Vectors;`

daviddesmet commented on test/NaCl.Core.SimdTests/Salsa20ScalarTests.cs:
```diff
@@ -0,0 +1,432 @@
+    using Vectors;
+    using System.Linq;
+    using NaCl.Core.Base.SalsaCore;
+    using NaCl.Core.SimdTests.Vectors;
```
I don't think this `using` is needed since above it you are already specifying `using Vectors;`

daviddesmet commented:
@TimothyMakkison my apologies for the late follow-up, it has been hectic over here.

Thanks a ton for pulling this big task off, I added a few comments to the PR; most of them are minor stuff.

I also noticed a PowerShell script but I didn't see it being called from DevOps, is it something you intended to do? The build defined in this project uses Cake, we could set the environment variable in there afaik; I will need to investigate further into that. Perhaps we can use that and just define tests that should be running with specific environment variables without the need to update the GitHub actions.

Last but not least, I wonder if you could work on the coverage a little bit more since we dropped quite a lot with this PR, you can check codecov from above.

daviddesmet commented on src/NaCl.Core/Base/ChaCha20BaseIntrinsics.cs:
```diff
@@ -0,0 +1,69 @@
+    public static unsafe void ChaCha20(Span<uint> state, ReadOnlySpan<byte> input, Span<byte> output, ulong bytes)
+    {
+        if (!Sse3.IsSupported || !BitConverter.IsLittleEndian)
+            throw new Exception("Error this vectorisation is not supported on this CPU");
```
We should avoid using the base `Exception` and instead use `NotSupportedException`, I would also suggest using a more specific error message like `The SSE3 vectorisation is not supported on the current CPU`.

daviddesmet commented on src/NaCl.Core/Base/ChaCha20BaseIntrinsics.cs:
```diff
@@ -0,0 +1,69 @@
+    public static unsafe void HChaCha20(ReadOnlySpan<uint> state, Span<byte> subKey)
+    {
+        if (!Sse3.IsSupported || !BitConverter.IsLittleEndian)
+            throw new Exception("Error this vectorisation is not supported on this CPU");
```
We should avoid using the base `Exception` and instead use `NotSupportedException`, I would also suggest using a more specific error message like `The SSE3 vectorisation is not supported on the current CPU`.

daviddesmet commented on src/NaCl.Core/Base/ChaCha20BaseIntrinsics.cs:
```diff
@@ -0,0 +1,69 @@
+    public static unsafe void ChaCha20KeyStream(ReadOnlySpan<uint> state, Span<byte> output)
+    {
+        if (!Sse3.IsSupported || !BitConverter.IsLittleEndian)
+            throw new Exception("Error this vectorisation is not supported on this CPU");
```
We should avoid using the base `Exception` and instead use `NotSupportedException`, I would also suggest using a more specific error message like `The SSE3 vectorisation is not supported on the current CPU`.

daviddesmet commented on src/NaCl.Core/Base/Salsa20BaseIntrinsics.cs:
```diff
@@ -0,0 +1,69 @@
+    public static unsafe void Salsa20(Span<uint> state, ReadOnlySpan<byte> input, Span<byte> output, ulong bytes)
+    {
+        if (!Sse3.IsSupported || !BitConverter.IsLittleEndian)
+            throw new Exception("Error this vectorisation is not supported on this CPU");
```
We should avoid using the base `Exception` and instead use `NotSupportedException`, I would also suggest using a more specific error message like `The SSE3 vectorisation is not supported on the current CPU`.

daviddesmet commented on src/NaCl.Core/Base/Salsa20BaseIntrinsics.cs:
```diff
@@ -0,0 +1,69 @@
+    public static unsafe void HSalsa20(ReadOnlySpan<uint> state, Span<byte> subKey)
+    {
+        if (!Sse3.IsSupported || !BitConverter.IsLittleEndian)
+            throw new Exception("Error this vectorisation is not supported on this CPU");
```
We should avoid using the base `Exception` and instead use `NotSupportedException`, I would also suggest using a more specific error message like `The SSE3 vectorisation is not supported on the current CPU`.

daviddesmet commented on src/NaCl.Core/Base/Salsa20BaseIntrinsics.cs:
```diff
@@ -0,0 +1,69 @@
+    public static unsafe void Salsa20KeyStream(ReadOnlySpan<uint> state, Span<byte> output)
+    {
+        if (!Sse3.IsSupported || !BitConverter.IsLittleEndian)
+            throw new Exception("Error this vectorisation is not supported on this CPU");
```
We should avoid using the base `Exception` and instead use `NotSupportedException`, I would also suggest using a more specific error message like `The SSE3 vectorisation is not supported on the current CPU`.

TimothyMakkison commented:
Thanks for the review. I got sidetracked and forgot about this PR so it's unfinished. 

> I wonder if we do need the preprocessor directive since we are checking if the hardware supports it

`System.Runtime.Intrinsics.X86` is only [supported for .NET 3.0+](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.intrinsics.x86.sse3?view=net-7.0#applies-to). Intrinsics code won't compile for libraries that multi target versions unless we use a preprocessor.


> Could you make this access modifier implicit?

Not sure what you mean by implicit, do you mean `private`?


> We could switch to scoped namespace declaration here since we already using it on the other files =)

Should I convert all the files in the solution to scoped namespaces?


> I also noticed a PowerShell script but I didn't see it being called from DevOps, is it something you intended to do? The build defined in this project uses Cake, we could set the environment variable in there afaik; 

Sorry I was using the benchmark code and powershell script for testing and forgot to remove them. I'll modify the cake file.


Do you have an opinion on the refactor? 

The logic up to 202ff15 was [fairly simple](https://github.com/daviddesmet/NaCl.Core/blob/202ff15f1fc0fa74e842cbe384581e490f1f42c1/src/NaCl.Core/Base/ChaCha20Base.cs); the intrinsics code was added with pre processing and an IsSupported check. This has the advantage of not touching the api, keeping the code/code flow the same and adding minimal additional code and files.

Commit 4d68791 separates the intrinsic and non intrinsic code from SalsaBase/ChaChaBase creating individual files for them, SalsaBase/ChaChaBase act as a facade [calling the internal IChaChaCore interface](https://github.com/daviddesmet/NaCl.Core/blob/4d68791ea70192c160e6afa640271530db5c7ccb/src/NaCl.Core/Base/ChaCha20Base.cs) for all methods. This has the disadvantage of adding a lot of code, altering a lot of code to keep the api the same, duplicating code and arguably being overly complex.

I'll probably undo the refactor unless you prefer it, I thinks it's too complicated, harder to maintain, and isn't needed if the CI will test SIMD and non SIMD versions. Should also improve the code coverage and reduce duplication.

TimothyMakkison commented:
Thanks for the review. I got sidetracked and forgot about this PR so it's unfinished. 

> I wonder if we do need the preprocessor directive since we are checking if the hardware supports it

`System.Runtime.Intrinsics.X86` is only [supported for .NET 3.0+](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.intrinsics.x86.sse3?view=net-7.0#applies-to). Intrinsics code won't compile for libraries that multi target versions unless we use a preprocessor.


> Could you make this access modifier implicit?

Not sure what you mean by implicit, do you mean `private`?


> We could switch to scoped namespace declaration here since we already using it on the other files =)

Should I convert all the files in the solution to scoped namespaces?


> I also noticed a PowerShell script but I didn't see it being called from DevOps, is it something you intended to do? The build defined in this project uses Cake, we could set the environment variable in there afaik; 

Sorry I was using the benchmark code and powershell script for testing and forgot to remove them. I'll modify the cake file.


Do you have an opinion on the refactor? 

The logic up to 202ff15 was [fairly simple](https://github.com/daviddesmet/NaCl.Core/blob/202ff15f1fc0fa74e842cbe384581e490f1f42c1/src/NaCl.Core/Base/ChaCha20Base.cs); the intrinsics code was added with pre processing and an IsSupported check. This has the advantage of not touching the api, keeping the code/code flow the same and adding minimal additional code and files.

Commit 4d68791 separates the intrinsic and non intrinsic code from SalsaBase/ChaChaBase creating individual files for them, SalsaBase/ChaChaBase act as a facade [calling the internal IChaChaCore interface](https://github.com/daviddesmet/NaCl.Core/blob/4d68791ea70192c160e6afa640271530db5c7ccb/src/NaCl.Core/Base/ChaCha20Base.cs) for all methods. This has the disadvantage of adding a lot of code, altering a lot of code to keep the api the same, duplicating code and arguably being overly complex.

I'll probably undo the refactor unless you prefer it, I thinks it's too complicated, harder to maintain, and isn't needed if the CI will test SIMD and non SIMD versions. Should also improve the code coverage and reduce duplication.

TimothyMakkison committed:
Remove unused code

TimothyMakkison committed:
Rename internal protected to protected internal

TimothyMakkison committed:
Remove unnecessary usings

TimothyMakkison committed:
Simplified slices

TimothyMakkison committed:
Removed unnecessary using statements

TimothyMakkison committed:
Add explcit access modifiers

TimothyMakkison committed:
Edit error message, move namespace and change visibility to internal

TimothyMakkison committed:
Remove internal core code, replaced with pre processor functions

TimothyMakkison committed:
Delete core scalar and itnrinsics tests

TimothyMakkison committed:
Code cleanup

TimothyMakkison committed:
Deleted powershell script and updated cake script

TimothyMakkison committed:
Correct visibility, remove unneeded InternalsVisibleTo, minor chnages

TimothyMakkison committed:
Add Salsa64 SSE41 shuffle optimisation, updated guards/checks, changed the cake file and added more comments.

TimothyMakkison committed:
Add Salsa64 SSE41 shuffle optimisation, updated guards/checks and added more comments.

TimothyMakkison commented:
> Yeah, I agree with that. It added a lot of maintainability complexity, if we limit the scope of the supported .NET versions we can stick with the first logic you proposed.
> I think we can drop .NET 3 and just stick with the supported versions going forward. Supporting obsolete versions of .NET just adds a lot of maintainability headaches.

The "refactor" was added to make the code testable, with the CI and cake file this thankfully isn't needed. This pr is compatible with all the different versoins of .NET it just requires a couple of pre processor checks.

- Updated the cake file, NET 7 RC 2 has broken my sdk so I can't install the cake tool and test it. Hopefully it works 🤞 
- Reverted the changes to `Salsa20Base` and `ChaCha20Base`.
- Deleted intrinsic/scalar tests.
- Optimized the shuffle function for `Salsa64`, this used a different instruction set so I updated the guard checks.
- To reduce the scope of this pr I haven't updated all the namespaces. I'm tempted to create a code cleanup pr changing the namespaces, exception parameters, empty arrays and unused fields etc.

TimothyMakkison committed:
Merge branch 'merge_intrinsics' of https://github.com/TimothyMakkison/NaCl.Core into merge_intrinsics

daviddesmet commented:
> * To reduce the scope of this PR I haven't updated all the namespaces. I'm tempted to create a code cleanup pr changing the namespaces, exception parameters, empty arrays and unused fields etc.

Yeah, sounds great!

> Do you think it would be worth getting someone who knows intrinsics to double check this? This was my first time using intrinisics/unsafe code in this way.

That would be nice but I really don't know anyone from my side. :/

TimothyMakkison commented:
I had the cunning idea of submitting a pr to the bouncycastle project. Give them a performance bump for large cipherstreams and have another set of eyes look at the implementations. Once I get a grasp of `IDigest` I'll make a pr
